tsl/src/chunk.c

```diff
@@ -259,3 +260,94 @@ chunk_invoke_drop_chunks(Oid relid, Datum older_than, Datum older_than_type)
 
 	return num_results;
 }
+
+/*
+ * drop_chunk_replica:
+ *
+ * This function drops a chunk on a specified data node. It then
+ * removes the metadata about the datanode, chunk association on the
+ * access node.
+ */
+Datum
+drop_chunk_replica(PG_FUNCTION_ARGS)
+{
+	const char *node_name = PG_ARGISNULL(0) ? NULL : NameStr(*PG_GETARG_NAME(0));
+	Oid chunk_id = PG_ARGISNULL(1) ? InvalidOid : PG_GETARG_OID(1);
+	List *data_nodes = NIL, *replicas;
+	ForeignServer *server;
+	Chunk *chunk;
+	ChunkDataNode *cdn;
+	char *drop_cmd;
+
+	if (!OidIsValid(chunk_id))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("Invalid chunk specification %u", chunk_id)));
+
+	chunk = ts_chunk_get_by_relid(chunk_id, false);
+
+	if (NULL == chunk)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("%u is not a valid chunk oid", chunk_id)));
+
+	/* It has to be a foreign table chunk */
+	if (chunk->relkind != RELKIND_FOREIGN_TABLE)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("\"%s\" is not a valid remote chunk", get_rel_name(chunk_id))));
+
+	server = data_node_get_foreign_server(node_name, ACL_USAGE, true, false);
+	Assert(NULL != server);
+
+	/* Early abort on missing permissions */
+	ts_hypertable_permissions_check(chunk_id, GetUserId());
+
+	/* check that the chunk is indeed present on the specified data node */
+	cdn = ts_chunk_data_node_scan_by_chunk_id_and_node_name(chunk->fd.id,
+															server->servername,
+															CurrentMemoryContext);
+
+	if (!cdn)
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("Chunk \"%s\" does not exist on data node \"%s\"",
+						get_rel_name(chunk_id),
+						node_name)));
+
+	/* Ensure that this chunk remains replicated so as to avoid data loss */
+	replicas = ts_chunk_data_node_scan_by_chunk_id(chunk->fd.id, CurrentMemoryContext);
+
+	/*
+	 * There should be atleast one surviving replica after the deletion here.
+	 *
+	 * We could fetch the corresponding hypertable and check its
+	 * replication_factor. But the user of this function is using it
+	 * to move chunk from one datanode to another and is well aware of
+	 * the replication_factor requirements
+	 */
+	if (list_length(replicas) <= 1)
+		ereport(ERROR,
+				(errcode(ERRCODE_TS_INSUFFICIENT_NUM_DATA_NODES),
+				 errmsg("dropping this chunk \"%s\" could lead to data loss",
+						get_rel_name(chunk_id))));
+
+	/*
+	 * Drop chunk on the data node using a regular "DROP TABLE". CASCADE is
+	 * not required and compressed chunk (if any) should also get dropped.
+	 */
+	drop_cmd = psprintf("DROP TABLE %s.%s",
+						quote_identifier(chunk->fd.schema_name.data),
+						quote_identifier(chunk->fd.table_name.data));
+	data_nodes = lappend(data_nodes, (char *) node_name);
```

Comment:
`lappend` is confusing here, since the list is supposed to contain only one node, right? I suggest to create the list with one element instead. E.g., with `make_list1`.

Reply:
Ok, will use `list_make1`. Thanks